### ISSUE TITLE
HSQLDB v2.7.2 incompatible

### ISSUE THREAD

**Issue by RohdeSchwarz-SDC** [spannm/ucanaccess#6]
Issue
--------------------
Currently UCanAccess depends on `HSQLDB v2.7.1`, but at some point we will have to update this dependency.

I updated HSQLDB to `v2.7.2`, to check if it runs stable, but it doesn't.

**→ Is there something we can do in the UCanAccess library to avoid the error (see below) or is it possibly a bug in HSQLDB?**

Error
--------------------
```
net.ucanaccess.jdbc.UcanaccessSQLException: UCAExc:::5.1.1 Cannot invoke "org.hsqldb.Statement.getTableNamesForRead()" because "<parameter1>.sessionContext.currentStatement" is null
    at net.ucanaccess.jdbc.UcanaccessSQLException.wrap(UcanaccessSQLException.java:123)
    at net.ucanaccess.util.Try.orThrow(Try.java:298)
    at net.ucanaccess.jdbc.UcanaccessStatement.tryCatch(UcanaccessStatement.java:406)
    at net.ucanaccess.jdbc.UcanaccessStatement.executeUpdate(UcanaccessStatement.java:165)
    at my.namespace.MyTest.TestUCanAccessWithHSQLDBv272(MyTest.java:*)
```

Reproducer
--------------------
```
@Test
public void TestUCanAccessWithHSQLDBv272() throws SQLException, IOException {
    try(var conn = openNewDatabase(); var stmt = conn.createStatement()) {
        // create new table and add some data
        stmt.executeUpdate("CREATE TABLE Tbl (ID INT NOT NULL PRIMARY KEY, Txt VARCHAR(255))");
        stmt.executeUpdate("INSERT INTO Tbl (ID, Txt) VALUES (1, 'One'), (2, 'Two'), (3, 'Tre')");
        
        // query data
        try(var row = stmt.executeQuery("SELECT ID, Txt FROM Tbl ORDER BY ID")) {
            assertTrue(row.next()); assertEquals(1, row.getInt(1)); assertEquals("One", row.getString(2));
            assertTrue(row.next()); assertEquals(2, row.getInt(1)); assertEquals("Two", row.getString(2));
            assertTrue(row.next()); assertEquals(3, row.getInt(1)); assertEquals("Tre", row.getString(2));
        }
        // update data
        stmt.executeUpdate("UPDATE Tbl SET Txt = 'Row ' || Txt");
        try(var row = stmt.executeQuery("SELECT Txt FROM Tbl ORDER BY ID")) {
            assertTrue(row.next()); assertEquals("Row One", row.getString(1));
            assertTrue(row.next()); assertEquals("Row Two", row.getString(1));
            assertTrue(row.next()); assertEquals("Row Tre", row.getString(1));
        }
        
        // remove data
        stmt.executeUpdate("DELETE FROM Tbl");
        try(var row = stmt.executeQuery("SELECT * FROM Tbl")) {
            assertFalse(row.next());    // no data
        }
    }
}

private static Connection openNewDatabase() throws SQLException, IOException {
    var tempDbFile = Path.of(System.getProperty("java.io.tmpdir"), "TestDatabase.accdb");
    var dbFileLocation = tempDbFile.toString().replace('\\', '/');
    
    tempDbFile.toFile().delete();    // delete from previous test run
    
    return DriverManager.getConnection("jdbc:ucanaccess://" + dbFileLocation + ";newdatabaseversion=V2016");
}
```

**Comment by spannm** [comment]
Hello @RohdeSchwarz-SDC,

I encountered this issue back in November when upgrading the project's `hsqldb` dependency from v2.5.0 to v2.7.1
(imo this upgrade was important to address [CVE-2022-41853](https://access.redhat.com/security/cve/cve-2022-41853)).
I had wanted to go to v2.7.2 at the time but ran into the below-mentioned issue.
Despite some research (diffing changes between v2.7.1 and v2.7.2 etc.) I could not pinpoint the exact source of the problem at the time however I also suspect a flaw in hsqldb v2.7.2

Luckily v2.7.1 seems sound and cve-free, as such there is no pressing need to upgrade asap.
The upgrade has been on my to-do list. I just didn't get around to investigating in depth.
If you manage to chase down the issue and even offer a patch to the hsqldb people that would of course be superb.
I am happy to keep this ticket open for tracking until resolved.

cheers
Markus

**Comment by spannm** [comment]
@RohdeSchwarz-SDC fyi,

Condensed your reproducing code hoping to narrow down the issue:
```
@Test
public void testWithHSQLDBv272Mini() throws Exception {
    try (Connection conn = openNewDatabase();
         Statement stmt = conn.createStatement()) {

        // create new table and add some data
        stmt.executeUpdate("CREATE TABLE Tbl (Txt VARCHAR(255) NOT NULL PRIMARY KEY)");
        stmt.executeUpdate("INSERT INTO Tbl (Txt) VALUES ('One')");

        // update data
        stmt.executeUpdate("UPDATE Tbl SET Txt = 'Row ' || Txt");
    }
}
```

Issue occurs during update, only if data is present in the table (i.e. there is something to update; remove the insert and no problem).
It seems connected with transaction logic in hsqldb.

NPE in `TransactionManagerCommon.endActionTPL(Session)` line 425:
```
    HsqlName[] readLocks =
            session.sessionContext.currentStatement.getTableNamesForRead();
```

The last prior write to `currentStatement` takes place in `org.hsqldb.Session` line 1332 where the member is explicitly set to null:
`sessionContext.currentStatement = null;`

**Comment by spannm** [comment]
Maven debug settings:

```
<dep.hsqldb.version>2.7.2</dep.hsqldb.version>
<dep.hsqldb.classifier>debug</dep.hsqldb.classifier>
```

**Comment by spannm** [comment]
UCanAccess makes use of various hsqldb triggers to maintain a column cache (amongst other things).
This approach worked up to and including hsqldb version 2.7.1

The triggers query database meta data of the table for which a trigger is called.
The query utilizes the thread's UCanAccess connection and underlying hsqldb connection.

In v2.7.1 `org.hsqldb.Session` `sessionContext.currentStatement` is set to null in line 1332.
Subsequent access to `currentStatement` in `TransactionManagerCommon.endActionTPL` line 425 causes a null pointer exception.

After disabling these triggers (for test) the issue does not appear.

While the problem is now understood, how and where to fix (hsqldb or UCanAccess) it is not immediately obvious.

**Comment by fredtoussi** [comment]
Please see my update at SourceForge.

**Comment by spannm** [comment]
HyperSQL bug #1697 fixed
 (see https://sourceforge.net/p/hsqldb/bugs/1697/)

Will release UCanAccess after release of Hsqldb v2.7.3

**Comment by spannm** [comment]
Just noticed that Hsqldb v2.7.3 was released in the mean time 🥳
Will upgrade Hsqldb in UCanAccess soon.

Maven Central Link: https://central.sonatype.com/artifact/org.hsqldb/hsqldb/versions

**Comment by spannm** [comment]
Minor issued encountered while upgrading to v2.7.3
[Column ordering of compound primary keys in database metadata](https://sourceforge.net/p/hsqldb/bugs/1711/)
But we can work around this in UCanAccess.

The latest snapshot now uses HSQLDB 2.7.3

**Comment by spannm** [comment]
Fixed in [release 5.1.2](https://github.com/spannm/ucanaccess/releases/tag/ucanaccess-5.1.2)
